File: 2182/2182.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

char * repeatLimitedString(char * s, int repeatLimit){
    // count all numbers
    int arr[26] = {};
    memset(arr, 0, sizeof(arr));
    int idx;
    for (idx = 0; s[idx] != '\0'; ++idx) {
        arr[s[idx]-'a']++;
    }
    char * newS = malloc(100001 * sizeof(char));
    int repeatTimes = 0;
    int now = 25;
    for (idx = 0; s[idx] != '\0'; ++idx) {
        while(now >= 0 && !arr[now]) {
            now--;
            repeatTimes = 0;
        }
        if (now < 0) {
            break;
        }
        //printf("%d\n", repeatTimes);
        if (repeatTimes >= repeatLimit) {
            int tmp = now-1;
            while(tmp >= 0 && !arr[tmp]) {
                tmp--;
            }
            if (tmp < 0) {
                break;
            }
            newS[idx] = 'a' + tmp;
            arr[tmp]--;
            repeatTimes = 0;
        } else {
            newS[idx] = 'a' + now;
            arr[now]--;
            repeatTimes++;
        }
    }
    newS[idx] = '\0';
    return newS;
}

int main() {
    char * str = "";
    char * nStr = repeatLimitedString(str, 2);
    printf("%s", nStr);
    free(nStr);
    return 0;
}

File: 2182/2182.rs
struct Solution {}

impl Solution {
    pub fn repeat_limited_string(s: String, repeat_limit: i32) -> String {
        let mut char_count: [i32; 26] = [0; 26];
        // Count char number
        for c in s.chars() {
            char_count[c as usize - 'a' as usize] += 1;
        }
        let mut idx: i32 = 25;
        let mut ans = String::new();
        let mut count: i32 = 0;
        while idx >= 0 {
            while idx >= 0 && char_count[idx as usize] == 0 {
                idx -= 1;
                count = 0;
            }
            if idx < 0 {
                break;
            }
            if count >= repeat_limit {
                // find next number to insert
                let mut tmp: i32 = idx - 1;
                while tmp >= 0 && char_count[tmp as usize] == 0 {
                    tmp -= 1;
                }
                if tmp < 0 {
                    break;
                }
                // insert char
                ans.push(('a' as u8 + tmp as u8) as char);
                char_count[tmp as usize] -= 1;
                count = 0;
                continue;
            }
            ans.push(('a' as u8 + idx as u8) as char);
            char_count[idx as usize] -= 1;
            count += 1;
        }
        ans
    }
}
